language/vm/src/deserializer.rs

```diff
@@ -947,34 +950,74 @@ fn load_signature_token(cursor: &mut VersionedCursor) -> BinaryLoaderResult<Sign
     }
 }
 
-fn load_nominal_resource_flag(cursor: &mut VersionedCursor) -> BinaryLoaderResult<bool> {
-    if let Ok(byte) = cursor.read_u8() {
-        Ok(match SerializedNominalResourceFlag::from_u8(byte)? {
-            SerializedNominalResourceFlag::NOMINAL_RESOURCE => true,
-            SerializedNominalResourceFlag::NORMAL_STRUCT => false,
-        })
-    } else {
-        Err(PartialVMError::new(StatusCode::MALFORMED).with_message("Unexpected EOF".to_string()))
-    }
+#[derive(Copy, Clone)]
+enum AbilitySetPosition {
+    FunctionTypeParameters,
+    StructTypeParameters,
+    StructModifiers,
 }
 
-fn load_kind(cursor: &mut VersionedCursor) -> BinaryLoaderResult<Kind> {
-    if let Ok(byte) = cursor.read_u8() {
-        Ok(match SerializedKind::from_u8(byte)? {
-            SerializedKind::ALL => Kind::All,
-            SerializedKind::COPYABLE => Kind::Copyable,
-            SerializedKind::RESOURCE => Kind::Resource,
-        })
-    } else {
-        Err(PartialVMError::new(StatusCode::MALFORMED).with_message("Unexpected EOF".to_string()))
+fn load_ability_set(
+    cursor: &mut VersionedCursor,
+    pos: AbilitySetPosition,
+) -> BinaryLoaderResult<AbilitySet> {
+    let byte = match cursor.read_u8() {
+        Ok(byte) => byte,
+        Err(_) => {
+            return Err(PartialVMError::new(StatusCode::MALFORMED)
+                .with_message("Unexpected EOF".to_string()))
+        }
+    };
+
+    if cursor.version() < 2 {
```

Comment:
A short comment explaining how we interpret the old values in terms of ability sets would be helpful.